src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs

```diff
@@ -22,93 +23,220 @@ public class VsTestRunner : ITestRunner
         {
-            TestRunResult testResult = new TestRunResult() { Success = false };
+            var envVars = new Dictionary<string, string>();
+            if (mutationId != null)
+            {
+                envVars["ActiveMutation"] = mutationId.ToString();
+            }
+
+            var testCases = (mutationId == null || !_flags.HasFlag(OptimizationFlags.CoverageBasedTest)) ? null : _coverage.GetTests<TestCase>(mutationId.Value);
+            return RunVsTest(testCases, timeoutMs, envVars);
+        }
 
-            var testResults = RunAllTests(activeMutationId, GenerateRunSettings(timeoutMS ?? 0));
+        private TestRunResult RunVsTest(ICollection<TestCase> testCases, int? timeoutMs,
+            Dictionary<string, string> envVars)
+        {
+            var testResults = RunAllTests(testCases, envVars, GenerateRunSettings(timeoutMs, false), false);
 
             // For now we need to throw an OperationCanceledException when a testrun has timed out. 
             // We know the testrun has timed out because we received less test results from the test run than there are test cases in the unit test project.
-            if (testResults.Count() < _testCasesDiscovered)
+            var resultAsArray = testResults as TestResult[] ?? testResults.ToArray();
+            if (resultAsArray.All(x => x.Outcome != TestOutcome.Failed) && resultAsArray.Count() < (testCases ?? _discoveredTests).Count)
             {
                 throw new OperationCanceledException();
             }
 
-            testResult = new TestRunResult
+            var testResult = new TestRunResult
             {
-                Success = testResults.All(tr => tr.Outcome == TestOutcome.Passed),
+                Success = resultAsArray.All(tr => tr.Outcome == TestOutcome.Passed || tr.Outcome == TestOutcome.Skipped),
                 ResultMessage = string.Join(
                     Environment.NewLine,
-                    testResults.Where(tr => !string.IsNullOrWhiteSpace(tr.ErrorMessage)).Select(tr => tr.ErrorMessage)),
-                TotalNumberOfTests = _testCasesDiscovered
+                    resultAsArray.Where(tr => !string.IsNullOrWhiteSpace(tr.ErrorMessage))
+                        .Select(tr => tr.ErrorMessage)),
+                TotalNumberOfTests = _discoveredTests.Count()
             };
 
             return testResult;
         }
 
-        public void DiscoverTests(string runSettings = null)
+        public TestRunResult CaptureCoverage()
         {
-            var waitHandle = new AutoResetEvent(false);
-            var handler = new DiscoveryEventHandler(waitHandle, _messages);
-            _vsTestConsole.DiscoverTests(_sources, runSettings ?? GenerateRunSettings(0), handler);
-
-            waitHandle.WaitOne();
+            Logger.LogDebug($"Capturing coverage.");
+            const int maxAttempts = 3;
+            IList<TestCase> failed = null;
+            for (var retryCount = 0; retryCount < maxAttempts; retryCount++)
```

Comment:
yes, I had issue with capture coverage when it failed randomly. Now that I have fixed several race conditions (mostly in the communicationchannel class) it looks more stable.
But I am not yet confident this works every time, so this is defensive

I let you decide if we keep it or not

Reply:
Well, it depends on if the retries help. If you think that when the first try fails, the second try could succeed I don't see the harm in having a few retries. But I don't see why the second try should succeed if the first one didn't. 

Unlike in our retry logic for compiling here nothing is changed in between the tries, so why should the second or third try succeed?

Reply:
fair point: it is because most of the issue I ran into where some form of race conditions between VsTest threading model and the channel transmission.
We can remove it altogether, if we never need it again, it will not be hard to rewrite

Reply:
Removed retrying altogether